### PR TITLE
Features/lol/improve error responses

### PR DESCRIPTION
**Pull request by lliss**
Related to:
https://github.com/ProboCI/probo-coordinator/pull/84
and
https://github.com/ProboCI/probo/pull/68
overlaps
https://github.com/ProboCI/probo-proxy/pull/4

\We want to have the proxy return better errors for humans in some cases.
Particularly, in the case of Probo SASS, we want the probo frontend to
be redirected to a nice error page. If however there is no front end and
the user is visiting the proxy URL via a browser, they should also see a
nice(r) error page. To handle this, configuration settings were added
that allow the proxy to be configured with a redirection URL for error
handling and a string of custom 404 error HTML. We inspect the request's
headers to determine if the response should be HTML (redirection or
HTML for a browser) or JSON (cli) or a plaintext response (no accept
headers sent).

### To test
 - `npm test` and ensure all tests pass.
 - Follow tests for https://github.com/ProboCI/probo/pull/68 to set up builds that are reaped or in progress.
 - Visit the proxy URLs for those builds via your browser
   - If you have a redirectionURL set you should be redirected
   - If you have no redirectionURL you should get an HTML response
 - Visit the proxy via curl with no special headers set, you should get a plaintext response
 - Visit the proxy via curl JSON accept headers set, you should get a JSON response

### CODE REVIEW

File: lib/config.js [hunk 1]
```diff
@@ -29,15 +36,16 @@ if (argv.config) {
...
+  argv.config.forEach(function(val) {
+    loader.add(path.resolve(val));
+  });
 }
 
 var setOptions = {};
 var key = null;
 for (key in yargs.argv) {
-  if (yargs.argv[key] !== undefined) {
+  if (!_.isUndefined(yargs.argv[key])) {
```
Comment (tortillaj) [inline lib/config.js]: It's better to use native javascript `hasOwnProperty` here: 

```javascript
for (key in yargs.argv) {
  if (yargs.argv.hasOwnProperty(key)) {
    setOptions[key] = yargs.argv[key];
  }
}
```

Using `hasOwnProperty` ensures that the object you're looping has the property directly, and it wasn't inherited from its prototype chain somehow. This is how you'd typically perform a safe for loop in JS. From what I can tell, lodash is doing type checking (like the original code was doing), which is not what this loop test should be doing anyway.

Lastly, I rarely see a good reason to introduce a new dependency when native JS can function.

Comment (lliss) [inline lib/config.js]: Our linting setup flags this.

Comment (lliss) [inline lib/config.js]: I actually feel almost the total opposite. Lodash has a bunch of good features that prevent problems that using native JS tends to introduce. I think it is always a bad Idea to be looping over keys in this manner anyway and would was simply trying to get this to pass our linting setup without rewriting it.

Comment (tortillaj) [inline lib/config.js]: We don't want to do type checking only, so you can't use the original if test, or this lodash if test you made.

Comment (tortillaj) [inline lib/config.js]: @lliss but the thing is, it's not passing our linting setup _correctly_. The point of the linting rule is to ensure that you properly check the object has the property you're looping over, and you aren't instead looping over an inherited property. The lodash function does not do that. The lodash function is doing the same check it was originally doing (i.e. simply a type check).

`hasOwnProperty` does exactly what the linting rule is guarding against. if you still want to use lodash for some reason, read about [_.has](https://lodash.com/docs#has). In this case, the lodash function doesn't "prevent problems" in using the native functions. It just expands on them.

File: .eslintrc.yml [hunk 2]
```diff
@@ -0,0 +1,175 @@
+rules:
```
Comment (tizzo) [inline .eslintrc.yml]: This should be using [the shared probo rules](https://github.com/ProboCI/eslint-config-probo) see the [probo implementation](https://github.com/ProboCI/probo/blob/master/.eslintrc).

Comment (lliss) [inline .eslintrc.yml]: Probo rules needed to be updated... https://github.com/ProboCI/eslint-config-probo/pull/1

File: lib/config.js (hunk 1 above)
Comment (lliss) [inline lib/config.js]: Adjusted as per your comments.

**Comment by tortillaj** [comment]
@tizzo LGTM

**Comment by lliss** [comment]
Still mergable... amazing.

**Pull request closed without merging**